central/image/service/service_impl.go

```diff
@@ -462,21 +471,42 @@ func (s *serviceImpl) GetImageVulnerabilitiesInternal(ctx context.Context, reque
 	return internalScanRespFromImage(img), nil
 }
 
-func (s *serviceImpl) acquireScanSemaphore() error {
-	if err := s.internalScanSemaphore.Acquire(concurrency.AsContext(concurrency.Timeout(maxSemaphoreWaitTime)), 1); err != nil {
-		// Aborted indicates the operation was aborted, typically due to a concurrency
-		// issues.  Clients should retry by default on Aborted.
-		s, err := status.New(codes.Aborted, err.Error()).WithDetails(&v1.ScanImageInternalResponseDetails_TooManyParallelScans{})
-		if pkgUtils.ShouldErr(err) == nil {
-			return s.Err()
+func (s *serviceImpl) acquireScanSemaphore(ctx context.Context) error {
+	semaphoreCtx, cancel := context.WithTimeout(ctx, maxSemaphoreWaitTime)
+	defer cancel()
+	if err := s.internalScanSemaphore.Acquire(semaphoreCtx, 1); err != nil {
+		wrappedErr := errors.Wrap(err, "acquiring scan semaphore")
+
+		// If the context was canceled, we do not want to indicate the client to retry.
+		if errors.Is(err, context.Canceled) {
+			return status.Error(codes.Canceled, wrappedErr.Error())
 		}
+
+		// Aborted indicates the operation was aborted, typically due to concurrency issues.
+		// Clients should retry by default on Aborted.
```

Comment:
*Should* retry or *do* retry? I tried to find something which says which status codes are automatically retried and the only one I can potentially find is `UNAVAILABLE`. Can you confirm if we do retry on `ABORTED`?

--

Edit: we do retry because of the `TooManyParallelScans`. I do wonder if we should have just done `UNAVAILABLE` from the beginning? I imagine the builtin mechanism is an exponential backoff, which is what we want, but I haven't confirmed. This is fine as-is

--

Edit 2: oh this comment was already here and just moved 😄